### PR TITLE
Properly update retry_last_ts when hitting the maximum retry interval

### LINKED ISSUE #16101: Federation sender started spamming requests
### Description

Two of my three federation senders are spamming two specific servers very quickly. Restarting them solved it temporarily, but they quickly started spamming the same two hosts again, possibly after sending an event to a room the hosts were in. After that I tried adding

```yaml
federation:
  destination_min_retry_interval: 10m
  destination_retry_multiplier: 5
  destination_max_retry_interval: 365d
```

to the config and restarting. After that it didn't start spamming again, but I'm not sure if it was the config or if it just wanted to be restarted twice, as it had updated the `destinations` table with a new `retry_last_ts` and `retry_interval` (the new interval being the maximum)

### Homeserver

maunium.net

### Synapse Version

1.90.0rc1

### Database

PostgreSQL

### Workers

Multiple workers

### Platform

Docker (custom image)

### Relevant log output

```shell
2023-08-11 08:25:42,906 - synapse.federation.sender.per_destination_queue - 399 - WARNING - federation_transaction_transmission_loop-26175- TX [matrix.buffadoo.nl] Received 401 response to transaction: 401: Unauthorized
2023-08-11 08:25:42,915 - synapse.replication.tcp.resource - 193 - INFO - replication_notifier-5959- Streaming: caches -> 18455011
2023-08-11 08:25:42,917 - synapse.federation.sender.per_destination_queue - 486 - INFO - federation_transaction_transmission_loop-26176- Catching up destination matrix.buffadoo.nl with 1 PDUs
2023-08-11 08:25:42,919 - synapse.federation.sender.per_destination_queue - 568 - INFO - federation_transaction_transmission_loop-26176- Catching up rooms to matrix.buffadoo.nl: '!ehXvUhWNASUkSLvAGP:matrix.org'
2023-08-11 08:25:42,919 - synapse.federation.sender.transaction_manager - 121 - INFO - federation_transaction_transmission_loop-26176- TX [matrix.buffadoo.nl] {1691741868425} Sending transaction [1691741868425], (PDUs: 1, EDUs: 0)
2023-08-11 08:25:42,956 - synapse.http.matrixfederationclient - 715 - INFO - federation_transaction_transmission_loop-26176- {PUT-O-15600} [matrix.buffadoo.nl] Got response headers: 401 Unauthorized
2023-08-11 08:25:42,956 - synapse.http.matrixfederationclient - 805 - WARNING - federation_transaction_transmission_loop-26176- {PUT-O-15600} [matrix.buffadoo.nl] Request failed: PUT matrix-federation://matrix.buffadoo.nl/_matrix/federation/v1/send/1691741868425: HttpResponseException('401: Unauthorized')
2023-08-11 08:25:42,956 - synapse.util.retryutils - 234 - INFO - federation_transaction_transmission_loop-26176- Connection to matrix.buffadoo.nl was unsuccessful (<class 'synapse.api.errors.HttpResponseException'>(401: Unauthorized)); backoff now 604800000
2023-08-11 08:25:42,957 - synapse.federation.sender.transaction_manager - 171 - INFO - federation_transaction_transmission_loop-26176- TX [matrix.buffadoo.nl] {1691741868425} got 401 response
2023-08-11 08:25:42,957 - synapse.federation.sender.per_destination_queue - 399 - WARNING - federation_transaction_transmission_loop-26176- TX [matrix.buffadoo.nl] Received 401 response to transaction: 401: Unauthorized
2023-08-11 08:25:42,958 - synapse.federation.sender - 979 - INFO - wake_destinations_needing_catchup-1- Destination matrix.holms.place has outstanding catch-up, waking up.
2023-08-11 08:25:42,960 - synapse.federation.sender.per_destination_queue - 486 - INFO - federation_transaction_transmission_loop-26177- Catching up destination matrix.fuzuki.dev with 1 PDUs
2023-08-11 08:25:42,963 - synapse.federation.sender.per_destination_queue - 568 - INFO - federation_transaction_transmission_loop-26177- Catching up rooms to matrix.fuzuki.dev: '!SMloEYlhCiqKwRLAgY:fachschaften.org'
2023-08-11 08:25:42,963 - synapse.federation.sender.transaction_manager - 121 - INFO - federation_transaction_transmission_loop-26177- TX [matrix.fuzuki.dev] {1691741868426} Sending transaction [1691741868426], (PDUs: 1, EDUs: 0)
2023-08-11 08:25:42,965 - synapse.replication.tcp.resource - 193 - INFO - replication_notifier-5960- Streaming: caches -> 18455012
2023-08-11 08:25:42,966 - synapse.federation.sender.per_destination_queue - 486 - INFO - federation_transaction_transmission_loop-26178- Catching up destination matrix.buffadoo.nl with 1 PDUs
2023-08-11 08:25:42,969 - synapse.federation.sender.per_destination_queue - 568 - INFO - federation_transaction_transmission_loop-26178- Catching up rooms to matrix.buffadoo.nl: '!ehXvUhWNASUkSLvAGP:matrix.org'
2023-08-11 08:25:42,969 - synapse.federation.sender.transaction_manager - 121 - INFO - federation_transaction_transmission_loop-26178- TX [matrix.buffadoo.nl] {1691741868427} Sending transaction [1691741868427], (PDUs: 1, EDUs: 0)
2023-08-11 08:25:43,015 - synapse.http.matrixfederationclient - 715 - INFO - federation_transaction_transmission_loop-26178- {PUT-O-15602} [matrix.buffadoo.nl] Got response headers: 401 Unauthorized
2023-08-11 08:25:43,015 - synapse.http.matrixfederationclient - 805 - WARNING - federation_transaction_transmission_loop-26178- {PUT-O-15602} [matrix.buffadoo.nl] Request failed: PUT matrix-federation://matrix.buffadoo.nl/_matrix/federation/v1/send/1691741868427: HttpResponseException('401: Unauthorized')
2023-08-11 08:25:43,015 - synapse.util.retryutils - 234 - INFO - federation_transaction_transmission_loop-26178- Connection to matrix.buffadoo.nl was unsuccessful (<class 'synapse.api.errors.HttpResponseException'>(401: Unauthorized)); backoff now 604800000
2023-08-11 08:25:43,016 - synapse.federation.sender.transaction_manager - 171 - INFO - federation_transaction_transmission_loop-26178- TX [matrix.buffadoo.nl] {1691741868427} got 401 response
2023-08-11 08:25:43,016 - synapse.federation.sender.per_destination_queue - 399 - WARNING - federation_transaction_transmission_loop-26178- TX [matrix.buffadoo.nl] Received 401 response to transaction: 401: Unauthorized
2023-08-11 08:25:43,022 - synapse.replication.tcp.resource - 193 - INFO - replication_notifier-5961- Streaming: caches -> 18455013
2023-08-11 08:25:43,025 - synapse.federation.sender.per_destination_queue - 486 - INFO - federation_transaction_transmission_loop-26179- Catching up destination matrix.buffadoo.nl with 1 PDUs
2023-08-11 08:25:43,027 - synapse.federation.sender.per_destination_queue - 568 - INFO - federation_transaction_transmission_loop-26179- Catching up rooms to matrix.buffadoo.nl: '!ehXvUhWNASUkSLvAGP:matrix.org'
2023-08-11 08:25:43,027 - synapse.federation.sender.transaction_manager - 121 - INFO - federation_transaction_transmission_loop-26179- TX [matrix.buffadoo.nl] {1691741868428} Sending transaction [1691741868428], (PDUs: 1, EDUs: 0)
2023-08-11 08:25:43,066 - synapse.http.matrixfederationclient - 715 - INFO - federation_transaction_transmission_loop-26179- {PUT-O-15603} [matrix.buffadoo.nl] Got response headers: 401 Unauthorized
2023-08-11 08:25:43,067 - synapse.http.matrixfederationclient - 805 - WARNING - federation_transaction_transmission_loop-26179- {PUT-O-15603} [matrix.buffadoo.nl] Request failed: PUT matrix-federation://matrix.buffadoo.nl/_matrix/federation/v1/send/1691741868428: HttpResponseException('401: Unauthorized')
2023-08-11 08:25:43,067 - synapse.util.retryutils - 234 - INFO - federation_transaction_transmission_loop-26179- Connection to matrix.buffadoo.nl was unsuccessful (<class 'synapse.api.errors.HttpResponseException'>(401: Unauthorized)); backoff now 604800000
2023-08-11 08:25:43,067 - synapse.federation.sender.transaction_manager - 171 - INFO - federation_transaction_transmission_loop-26179- TX [matrix.buffadoo.nl] {1691741868428} got 401 response
2023-08-11 08:25:43,067 - synapse.federation.sender.per_destination_queue - 399 - WARNING - federation_transaction_transmission_loop-26179- TX [matrix.buffadoo.nl] Received 401 response to transaction: 401: Unauthorized
2023-08-11 08:25:43,073 - synapse.replication.tcp.resource - 193 - INFO - replication_notifier-5962- Streaming: caches -> 18455014
2023-08-11 08:25:43,075 - synapse.federation.sender.per_destination_queue - 486 - INFO - federation_transaction_transmission_loop-26180- Catching up destination matrix.buffadoo.nl with 1 PDUs
2023-08-11 08:25:43,077 - synapse.federation.sender.per_destination_queue - 568 - INFO - federation_transaction_transmission_loop-26180- Catching up rooms to matrix.buffadoo.nl: '!ehXvUhWNASUkSLvAGP:matrix.org'
2023-08-11 08:25:43,078 - synapse.federation.sender.transaction_manager - 121 - INFO - federation_transaction_transmission_loop-26180- TX [matrix.buffadoo.nl] {1691741868429} Sending transaction [1691741868429], (PDUs: 1, EDUs: 0)
2023-08-11 08:25:43,114 - synapse.http.matrixfederationclient - 715 - INFO - federation_transaction_transmission_loop-26180- {PUT-O-15604} [matrix.buffadoo.nl] Got response headers: 401 Unauthorized
2023-08-11 08:25:43,115 - synapse.http.matrixfederationclient - 805 - WARNING - federation_transaction_transmission_loop-26180- {PUT-O-15604} [matrix.buffadoo.nl] Request failed: PUT matrix-federation://matrix.buffadoo.nl/_matrix/federation/v1/send/1691741868429: HttpResponseException('401: Unauthorized')
2023-08-11 08:25:43,115 - synapse.util.retryutils - 234 - INFO - federation_transaction_transmission_loop-26180- Connection to matrix.buffadoo.nl was unsuccessful (<class 'synapse.api.errors.HttpResponseException'>(401: Unauthorized)); backoff now 604800000
2023-08-11 08:25:43,115 - synapse.federation.sender.transaction_manager - 171 - INFO - federation_transaction_transmission_loop-26180- TX [matrix.buffadoo.nl] {1691741868429} got 401 response
2023-08-11 08:25:43,115 - synapse.federation.sender.per_destination_queue - 399 - WARNING - federation_transaction_transmission_loop-26180- TX [matrix.buffadoo.nl] Received 401 response to transaction: 401: Unauthorized
2023-08-11 08:25:43,121 - synapse.http.matrixfederationclient - 715 - INFO - federation_transaction_transmission_loop-26177- {PUT-O-15601} [matrix.fuzuki.dev] Got response headers: 400 Bad Request
2023-08-11 08:25:43,121 - synapse.http.matrixfederationclient - 805 - WARNING - federation_transaction_transmission_loop-26177- {PUT-O-15601} [matrix.fuzuki.dev] Request failed: PUT matrix-federation://matrix.fuzuki.dev/_matrix/federation/v1/send/1691741868426: HttpResponseException('400: Bad Request')
2023-08-11 08:25:43,122 - synapse.federation.sender.transaction_manager - 171 - INFO - federation_transaction_transmission_loop-26177- TX [matrix.fuzuki.dev] {1691741868426} got 400 response
2023-08-11 08:25:43,122 - synapse.federation.sender.per_destination_queue - 399 - WARNING - federation_transaction_transmission_loop-26177- TX [matrix.fuzuki.dev] Received 400 response to transaction: 400: Bad Request
2023-08-11 08:25:43,122 - synapse.replication.tcp.resource - 193 - INFO - replication_notifier-5963- Streaming: caches -> 18455015
2023-08-11 08:25:43,124 - synapse.federation.sender.per_destination_queue - 486 - INFO - federation_transaction_transmission_loop-26181- Catching up destination matrix.buffadoo.nl with 1 PDUs
2023-08-11 08:25:43,127 - synapse.federation.sender.per_destination_queue - 568 - INFO - federation_transaction_transmission_loop-26181- Catching up rooms to matrix.buffadoo.nl: '!ehXvUhWNASUkSLvAGP:matrix.org'
2023-08-11 08:25:43,127 - synapse.federation.sender.transaction_manager - 121 - INFO - federation_transaction_transmission_loop-26181- TX [matrix.buffadoo.nl] {1691741868430} Sending transaction [1691741868430], (PDUs: 1, EDUs: 0)
2023-08-11 08:25:43,162 - synapse.http.matrixfederationclient - 715 - INFO - federation_transaction_transmission_loop-26181- {PUT-O-15605} [matrix.buffadoo.nl] Got response headers: 401 Unauthorized
2023-08-11 08:25:43,162 - synapse.http.matrixfederationclient - 805 - WARNING - federation_transaction_transmission_loop-26181- {PUT-O-15605} [matrix.buffadoo.nl] Request failed: PUT matrix-federation://matrix.buffadoo.nl/_matrix/federation/v1/send/1691741868430: HttpResponseException('401: Unauthorized')
2023-08-11 08:25:43,162 - synapse.util.retryutils - 234 - INFO - federation_transaction_transmission_loop-26181- Connection to matrix.buffadoo.nl was unsuccessful (<class 'synapse.api.errors.HttpResponseException'>(401: Unauthorized)); backoff now 604800000
2023-08-11 08:25:43,163 - synapse.federation.sender.transaction_manager - 171 - INFO - federation_transaction_transmission_loop-26181- TX [matrix.buffadoo.nl] {1691741868430} got 401 response
2023-08-11 08:25:43,163 - synapse.federation.sender.per_destination_queue - 399 - WARNING - federation_transaction_transmission_loop-26181- TX [matrix.buffadoo.nl] Received 401 response to transaction: 401: Unauthorized
2023-08-11 08:25:43,169 - synapse.replication.tcp.resource - 193 - INFO - replication_notifier-5964- Streaming: caches -> 18455016
2023-08-11 08:25:43,171 - synapse.federation.sender.per_destination_queue - 486 - INFO - federation_transaction_transmission_loop-26182- Catching up destination matrix.buffadoo.nl with 1 PDUs
2023-08-11 08:25:43,173 - synapse.federation.sender.per_destination_queue - 568 - INFO - federation_transaction_transmission_loop-26182- Catching up rooms to matrix.buffadoo.nl: '!ehXvUhWNASUkSLvAGP:matrix.org'
2023-08-11 08:25:43,173 - synapse.federation.sender.transaction_manager - 121 - INFO - federation_transaction_transmission_loop-26182- TX [matrix.buffadoo.nl] {1691741868431} Sending transaction [1691741868431], (PDUs: 1, EDUs: 0)
2023-08-11 08:25:43,213 - synapse.http.matrixfederationclient - 715 - INFO - federation_transaction_transmission_loop-26182- {PUT-O-15606} [matrix.buffadoo.nl] Got response headers: 401 Unauthorized
2023-08-11 08:25:43,213 - synapse.http.matrixfederationclient - 805 - WARNING - federation_transaction_transmission_loop-26182- {PUT-O-15606} [matrix.buffadoo.nl] Request failed: PUT matrix-federation://matrix.buffadoo.nl/_matrix/federation/v1/send/1691741868431: HttpResponseException('401: Unauthorized')
```

image

### PR DESCRIPTION
**Pull request by deepbluev7**
This was broken in 1.87 when the maximum retry interval got changed from almost infinite to a week (and made configurable).

fixes #16101

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [ ] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by deepbluev7** [comment]
Basically the problem is that if the retry ts is older than 8 days, but the retry interval is already 7 days, we would never update the last retry. This solution might cause a few too many writes, but those should only happen if we retry for other reasons, I hope?

**Comment by deepbluev7** [comment]
Actually, converting this to a draft, since this does not work, if the retry interval in the db is bigger than the new max retry interval and the server never was online in between...

**Comment by deepbluev7** [comment]
Okay, I think this works, but there where clauses confuse me :D

**Comment by deepbluev7** [comment]
Grafana dash after those changes at around 12:35:

image

File: changelog.d/16156.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix a bug introduced in 1.87 where synapse would send an excessive amount of federation requests to servers which have been offline for a long time. Contributed by Nico.
```
Comment (erikjohnston) [inline changelog.d/16156.bugfix:1]: This says 1.87 but the description says 1.90?

Comment (nico-famedly) [reply changelog.d/16156.bugfix:1]: I forgot to update the description, I somehow thought https://github.com/matrix-org/synapse/pull/15783 got merged into 1.90 when I wrote the commit message and then later fixed it.

**Comment by erikjohnston** [comment]
Ah, didn't spot that the fix has changed. I think the current fix does work, but equally for clarity I think we should add a clause to the upsert to ensure that we do update the `last_retry_ts`.

**Pull request merged**